// react-email/dist/source/utils/constants.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PACKAGE_PUBLIC_PATH = exports.PACKAGE_EMAILS_PATH = exports.EVENT_FILE_DELETED = exports.REACT_EMAIL_ROOT = exports.USER_STATIC_FILES = exports.USER_PACKAGE_JSON = exports.CURRENT_PATH = exports.PACKAGE_NAME = exports.DEFAULT_EMAILS_DIRECTORY = void 0;
const path_1 = __importDefault(require("path"));
// Package variables
exports.DEFAULT_EMAILS_DIRECTORY = 'emails';
exports.PACKAGE_NAME = 'react-email';
// Default paths
exports.CURRENT_PATH = process.cwd();
// User paths
exports.USER_PACKAGE_JSON = path_1.default.join(exports.CURRENT_PATH, 'package.json');
exports.USER_STATIC_FILES = path_1.default.join(exports.CURRENT_PATH, 'emails', 'static');
// React Email paths
exports.REACT_EMAIL_ROOT = path_1.default.join(exports.CURRENT_PATH, '.react-email');
// Events
exports.EVENT_FILE_DELETED = 'unlink';
exports.PACKAGE_EMAILS_PATH = path_1.default.join(exports.REACT_EMAIL_ROOT, exports.DEFAULT_EMAILS_DIRECTORY);
exports.PACKAGE_PUBLIC_PATH = path_1.default.join(exports.REACT_EMAIL_ROOT, 'public');
