// @react-email/img/dist/index.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  Img: () => Img
});
module.exports = __toCommonJS(src_exports);

// src/img.tsx
var React = __toESM(require("react"));
var import_jsx_runtime = require("react/jsx-runtime");
var Img = React.forwardRef(
  ({ alt, src, width, height, style, ...props }, forwardedRef) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "img",
    {
      ...props,
      ref: forwardedRef,
      "data-id": "react-email-img",
      alt,
      src,
      width,
      height,
      style: {
        display: "block",
        outline: "none",
        border: "none",
        textDecoration: "none",
        ...style
      }
    }
  )
);
Img.displayName = "Img";
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Img
});
