// react-email/dist/source/commands/export.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.exportTemplates = void 0;
const glob_1 = require("glob");
const esbuild_1 = __importDefault(require("esbuild"));
const tree_node_cli_1 = __importDefault(require("tree-node-cli"));
const ora_1 = __importDefault(require("ora"));
const log_symbols_1 = __importDefault(require("log-symbols"));
const render_1 = require("@react-email/render");
const fs_1 = require("fs");
const normalize_path_1 = __importDefault(require("normalize-path"));
const path_1 = __importDefault(require("path"));
const shelljs_1 = __importDefault(require("shelljs"));
const fs_2 = __importDefault(require("fs"));
const close_ora_on_sigint_1 = require("../utils/close-ora-on-sigint");
/*
  This first builds all the templates using esbuild and then puts the output in the `.js`
  files. Then these `.js` files are imported dynamically and rendered to `.html` files
  using the `render` function.
 */
const exportTemplates = async (outDir, srcDir, options) => {
    const spinner = (0, ora_1.default)('Preparing files...\n').start();
    (0, close_ora_on_sigint_1.closeOraOnSIGNIT)(spinner);
    const allTemplates = glob_1.glob.sync((0, normalize_path_1.default)(path_1.default.join(srcDir, '*.{tsx,jsx}')));
    esbuild_1.default.buildSync({
        bundle: true,
        entryPoints: allTemplates,
        platform: 'node',
        write: true,
        outdir: outDir,
    });
    const allBuiltTemplates = glob_1.glob.sync((0, normalize_path_1.default)(`${outDir}/*.js`), {
        absolute: true,
    });
    for (const template of allBuiltTemplates) {
        const component = await Promise.resolve(`${template}`).then(s => __importStar(require(s)));
        const rendered = (0, render_1.render)(component.default({}), options);
        const htmlPath = template.replace('.js', options.plainText ? '.txt' : '.html');
        (0, fs_1.writeFileSync)(htmlPath, rendered);
        (0, fs_1.unlinkSync)(template);
    }
    const staticDir = path_1.default.join(srcDir, 'static');
    const hasStaticDirectory = fs_2.default.existsSync(staticDir);
    if (hasStaticDirectory) {
        const result = shelljs_1.default.cp('-r', staticDir, path_1.default.join(outDir, 'static'));
        if (result.code > 0) {
            throw new Error(`Something went wrong while copying the file to ${outDir}/static, ${result.cat()}`);
        }
    }
    const fileTree = (0, tree_node_cli_1.default)(outDir, {
        allFiles: true,
        maxDepth: 4,
    });
    console.log(fileTree);
    spinner.stopAndPersist({
        symbol: log_symbols_1.default.success,
        text: 'Successfully exported emails',
    });
    process.exit();
};
exports.exportTemplates = exportTemplates;
